// @dynamic-mapper/angular/fesm5/dynamic-mapper-angular.js
import { InjectionToken, Injectable, Inject, NgModule } from '@angular/core';
import { __extends } from 'tslib';
import { MapperConfiguration, Mapper as Mapper$1 } from '@dynamic-mapper/mapper';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var MAPPING_PROFILE = new InjectionToken('MappingProfile');
var Mapper = /** @class */ (function (_super) {
    __extends(Mapper, _super);
    function Mapper(profiles) {
        return _super.call(this, new MapperConfiguration((/**
         * @param {?} cfg
         * @return {?}
         */
        function (cfg) { return profiles.forEach((/**
         * @param {?} p
         * @return {?}
         */
        function (p) { return cfg.addProfile(p); })); }))) || this;
    }
    Mapper.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    Mapper.ctorParameters = function () { return [
        { type: Array, decorators: [{ type: Inject, args: [MAPPING_PROFILE,] }] }
    ]; };
    return Mapper;
}(Mapper$1));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} profile
 * @return {?}
 */
function createProfileProvider(profile) {
    return { provide: MAPPING_PROFILE, useClass: profile, multi: true };
}
var MapperModule = /** @class */ (function () {
    function MapperModule() {
    }
    /**
     * @param {?} profiles
     * @return {?}
     */
    MapperModule.withProfiles = /**
     * @param {?} profiles
     * @return {?}
     */
    function (profiles) {
        return {
            ngModule: MapperModule,
            providers: profiles.map(createProfileProvider)
        };
    };
    MapperModule.decorators = [
        { type: NgModule, args: [{
                    providers: [Mapper]
                },] }
    ];
    return MapperModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { MAPPING_PROFILE, Mapper, MapperModule, createProfileProvider };
