// @dynamic-mapper/angular/fesm2015/dynamic-mapper-angular.js
import { InjectionToken, Injectable, Inject, NgModule } from '@angular/core';
import { Mapper as Mapper$1, MapperConfiguration } from '@dynamic-mapper/mapper';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MAPPING_PROFILE = new InjectionToken('MappingProfile');
class Mapper extends Mapper$1 {
    /**
     * @param {?} profiles
     */
    constructor(profiles) {
        super(new MapperConfiguration((/**
         * @param {?} cfg
         * @return {?}
         */
        cfg => profiles.forEach((/**
         * @param {?} p
         * @return {?}
         */
        p => cfg.addProfile(p))))));
    }
}
Mapper.decorators = [
    { type: Injectable }
];
/** @nocollapse */
Mapper.ctorParameters = () => [
    { type: Array, decorators: [{ type: Inject, args: [MAPPING_PROFILE,] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} profile
 * @return {?}
 */
function createProfileProvider(profile) {
    return { provide: MAPPING_PROFILE, useClass: profile, multi: true };
}
class MapperModule {
    /**
     * @param {?} profiles
     * @return {?}
     */
    static withProfiles(profiles) {
        return {
            ngModule: MapperModule,
            providers: profiles.map(createProfileProvider)
        };
    }
}
MapperModule.decorators = [
    { type: NgModule, args: [{
                providers: [Mapper]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { MAPPING_PROFILE, Mapper, MapperModule, createProfileProvider };
