// @dynamic-mapper/angular/__ivy_ngcc__/fesm2015/dynamic-mapper-angular.js
import { InjectionToken, Injectable, Inject, NgModule } from '@angular/core';
import { Mapper as Mapper$1, MapperConfiguration } from '@dynamic-mapper/mapper';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
import * as ɵngcc0 from '@angular/core';
const MAPPING_PROFILE = new InjectionToken('MappingProfile');
class Mapper extends Mapper$1 {
    /**
     * @param {?} profiles
     */
    constructor(profiles) {
        super(new MapperConfiguration((/**
         * @param {?} cfg
         * @return {?}
         */
        cfg => profiles.forEach((/**
         * @param {?} p
         * @return {?}
         */
        p => cfg.addProfile(p))))));
    }
}
Mapper.ɵfac = function Mapper_Factory(t) { return new (t || Mapper)(ɵngcc0.ɵɵinject(MAPPING_PROFILE)); };
Mapper.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: Mapper, factory: Mapper.ɵfac });
/** @nocollapse */
Mapper.ctorParameters = () => [
    { type: Array, decorators: [{ type: Inject, args: [MAPPING_PROFILE,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Mapper, [{
        type: Injectable
    }], function () { return [{ type: Array, decorators: [{
                type: Inject,
                args: [MAPPING_PROFILE]
            }] }]; }, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} profile
 * @return {?}
 */
function createProfileProvider(profile) {
    return { provide: MAPPING_PROFILE, useClass: profile, multi: true };
}
class MapperModule {
    /**
     * @param {?} profiles
     * @return {?}
     */
    static withProfiles(profiles) {
        return {
            ngModule: MapperModule,
            providers: profiles.map(createProfileProvider)
        };
    }
}
MapperModule.ɵfac = function MapperModule_Factory(t) { return new (t || MapperModule)(); };
MapperModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: MapperModule });
MapperModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [Mapper] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MapperModule, [{
        type: NgModule,
        args: [{
                providers: [Mapper]
            }]
    }], null, null); })();

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { MAPPING_PROFILE, Mapper, MapperModule, createProfileProvider };
